Do not allow IdentityItems with empty ids (resolves #100)

kevinlind opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Do not allow IdentityItems with empty ids.

- If an IdentityItem is created with an empty or null id, throw IllegalArgumentException from constructor.
- Add tests to validate reading JSON Identity Maps which contain items with null/empty ids are correctly removed when the IdentityMap object is created.

See Issue #100 

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

addb commented on code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:
```diff
@@ -49,8 +50,8 @@ public IdentityItem(
 		@Nullable final AuthenticatedState authenticatedState,
 		final boolean primary
 	) {
```
Update doc:
`@throws IllegalArgumentException if id is null or empty`

addb commented on code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:
```diff
@@ -193,7 +194,15 @@ static IdentityItem fromData(final Map<String, Object> data) {
 			return new IdentityItem(id, authenticatedState, primary);
```
Update doc:
`@throws IllegalArgumentException if id is null or empty`

addb reviewed (approved):
Left 1 comment for doc update. Apart from that looks good!

emdobrin commented on code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:
```diff
@@ -42,15 +43,15 @@ public final class IdentityItem {
-		if (id == null) {
-			throw new IllegalArgumentException("id must be non-null");
+		if (StringUtils.isNullOrEmpty(id)) {
+			throw new IllegalArgumentException("id must be non-null and non-empty");
```
is there a way to avoid throwing an exception for empty id value? I am thinking this could potentially break an existing implementation which erroneously sends empty ids

emdobrin commented on code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityMapTests.java:
```diff
@@ -310,6 +310,83 @@ public void test_FromData() throws Exception {
 	}
 
+	@Test
+	public void test_FromData_removesItem_emptyId() throws Exception {
```
Are we also covering the use-case where there may be previously stored ids with empty id value, so that the SDK clears them and stores the updated identity map? And if so, do we need a test for this scenario?

emdobrin commented on code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:
```diff
@@ -42,15 +43,15 @@ public final class IdentityItem {
 	 * @param id                 id for the item; should not be null
```
id for the item; should not be null or empty 

Also, don't forget to update the api reference page if we don't already cover this

kevinlind commented:
Hi @addb and @emdobrin . I've updated to code to ignore IdentityItems with empty ids instead of throwing an exception. Changing the IdentityItem API to throw is a breaking change as Emilia pointed out. I've added some functional tests to validate existing items with empty ids are not loaded on bootup. I also updated the documentation to state IdentityItems with empty ids are ignored when adding to an IdentityMap.

addb commented:
> Hi @addb and @emdobrin . I've updated to code to ignore IdentityItems with empty ids instead of throwing an exception. Changing the IdentityItem API to throw is a breaking change as Emilia pointed out. I've added some functional tests to validate existing items with empty ids are not loaded on bootup. I also updated the documentation to state IdentityItems with empty ids are ignored when adding to an IdentityMap.

I am not sure if this is the right approach as now we have 3 behaviors
|id|outcome|
|--|--|
|null | throws exception|
| emty| ignores|
|non-empty non-null String| accepted|

We should have similar behavior for error case, either we throw exceptions for both or ignore both

emdobrin commented on Documentation/api-reference.md:
```diff
@@ -432,7 +432,7 @@ Identity.updateIdentities(identityMap)
 ### IdentityMap
 
-Defines a map containing a set of end user identities, keyed on either namespace integration code or the namespace ID of the identity. The values of the map are an array, meaning that more than one identity of each namespace may be carried.
+Defines a map containing a set of end user identities, keyed on either namespace integration code or the namespace ID of the identity. The values of the map are an array, meaning that more than one identity of each namespace may be carried. Identities, represented by [IdentityItem](#identityitem) objects, may not have null or empty identitifers and are ignored when adding to the IdentityMap.
```
Defines a map containing a set of end user identities, keyed on either namespace integration code or the namespace ID of the identity. The values of the map are an array of [IdentityItem](#identityitem)(s), meaning that more than one identity of each namespace may be carried. Each IdentityItem should have a valid, non-null and non-empty identifier, otherwise it will be ignored.

emdobrin commented on code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:
```diff
@@ -64,6 +67,7 @@ public IdentityItem(
 	 * (@code primary} is set to false
 	 *
 	 * @param id the id for this {@link IdentityItem}; should not be null
+	 * @throws IllegalArgumentException if id is null or empty
```
if {@code id} is null

addb commented on code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:
```diff
@@ -37,12 +37,15 @@ public final class IdentityItem {
+	 * @throws IllegalArgumentException if {@code id} is null
 	 */
 	public IdentityItem(
 		@NonNull final String id,
```
`@NotBlank`
https://javaee.github.io/javaee-spec/javadocs/javax/validation/constraints/NotBlank.html

kevinlind replied:
This is not a standard AndroidX annotation, but instead from Hibernate-Validator in Java EE, so it's not really available to us.

addb commented on code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:
```diff
@@ -64,6 +67,7 @@ public IdentityItem(
 	 * @param id the id for this {@link IdentityItem}; should not be null
+	 * @throws IllegalArgumentException if id is null or empty
 	 */
 	public IdentityItem(@NonNull final String id) {
```
@NotBlank
https://javaee.github.io/javaee-spec/javadocs/javax/validation/constraints/NotBlank.html
